### DISCUSSION TITLE
thread/volatile

### DISCUSSION THREAD

**Discussion by giscus** [itwanger/tobebetterjavaer-giscus#248] [Announcements]
# thread/volatile

推荐阅读 学习路线👉 : 一份涵盖 Java、MySQL、Redis、C 语言、C++、Python、Go、操作系统、前端、数据结构与算法、蓝桥杯、大数据、Android、.NET的全方位编程学习路线！清晰且有效！ 面渣逆袭📗 ：面试前必刷，硬核理解版八股文，包括 Java 基础（JavaSE）、Java 集合框架、Java 并发编程（Java 多线程）、Java 虚拟机（JVM）、Spring、Redis、MyBatis、MySQL、操作系统、计算机网络、RocketMQ、分布式等等，助你拿到心仪 offer！ Java程序员常读书单📚：超1000本PDF，助力Java 程序员构建最强知识体系，涵盖Java、设计模式、数据库、数据结构与算法、大数据、架构、其他编程语言等等，再也不用辛苦去找下载地址了，这里全都有。 技术派实战项目 ：⭐️一款好用又强大的开源社区，附详细教程，包括Java、Spring、MySQL、Redis、微服务&分布式、消息队列、操作系统、计算机网络、数据结构与算法等计算机专业核心知识点。学编程，就上技术派😁。

https://javabetter.cn/thread/volatile.html

**Comment by hengxingstu** [comment]
volatile的关键是要记住这句话：
> 当程序执行到 volatile 变量的读操作或者写操作时，在其前面操作的更改肯定已经全部进行，且结果对后面的操作可见；在其后面的操作肯定还没有进行

volatile可以理解为一个特殊的时间点，所有之前的修改已经完成，之后的修改和读取也未开始。
还要谨记一个缺陷，**volatile并不提供原子性**，所以volatile变量的修改如果涉及多线程，必须要上锁。

**Comment by YoruWong** [comment]
> - 线程 A 执行到 `if (m_penguin == null)`，判断为 true，进入同步块。
> - 线程 B 执行到 `if (m_penguin == null)`，判断为 true，进入同步块。
>
> 如果线程 A 执行 `m_penguin = new Penguin()` 时发生指令重排序：
>
> - 线程 A 分配内存并设置引用，但尚未调用构造方法完成初始化。
> - 线程 B 此时判断 `m_penguin != null`，直接返回这个“半初始化”的对象。

这里貌似有些问题，在线程 A 设置引用时，线程 B 执行第一次判断，此时才会返回半初始化对象。若按原文中描述，线程 B 进入同步块时，线程 A 已经执行完同步块中的指令，m_penguin 已被完全初始化，故应为：

> - 线程 A 执行到 `if (m_penguin == null)`，判断为 true，进入同步块。
>
> 如果线程 A 执行 `m_penguin = new Penguin()` 时发生指令重排序：
>
> - 线程 A 分配内存并设置引用，但尚未调用构造方法完成初始化。
> - 线程 B 执行到 `if (m_penguin == null)`，判断为 false，直接返回这个“半初始化”的对象。

**Reply by Liucc-123** [reply]
你的理解是对的，应该是线程A半初始化对象后，此时，另一个线程B进来，判断对象不为空，返回这个半初始化的对象。

**Comment by MMA-JAY** [comment]
小王还是差点意思

**Comment by LFAW233** [comment]
这边提一嘴，在单例模式的代码中如果只关注功能实现，实际上同步代码块外的if判断是不必要的。
但一般开发模式中都会写上这个判断，避免锁竞争锁升级带来的额外开销。
